### PR TITLE
Core/AI: refactor SpellHit and SpellHitTarget.

### PR DESCRIPTION
**Pull request by ForesterDev**
**Changes proposed:**

- now caster/target is WorldObject instead of Unit
- remove SpellHitByGameObject / SpellHitTargetGameObject (handled by SpellHit / SpellHitTarget)
- rename parameters in scripts according parent methods

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Tests performed:** (Does it build, tested in-game, etc.)
Build

### CODE REVIEW

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_algalon_the_observer.cpp [hunk 1]
```diff
@@ -765,14 +765,18 @@ struct npc_living_constellation : public CreatureAI
         }
     }
 
-    void SpellHit(Unit* caster, SpellInfo const* spell) override
+    void SpellHit(WorldObject* caster, SpellInfo const* spellInfo) override
     {
-        if (spell->Id != SPELL_CONSTELLATION_PHASE_EFFECT || caster->GetTypeId() != TYPEID_UNIT)
+        Unit* unitCaster = caster->ToUnit();
```
Comment (Shauren) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_algalon_the_observer.cpp:770]: This changes logic - `caster->GetTypeId() != TYPEID_UNIT` excludes players, `ToUnit()` does not

**Comment by Shauren** [comment]
Generally this is a step in the right direction but I wonder if its correct for all scripts

example: script S casts spell X that hits creatures and gameobjects and has SpellHitTarget (old) that does something (to itself, not to targeted creature, think commented out target argument) if that spell hit a creature. Now after applying this change, it will also fire when hitting the gameobject

**Comment by Faq** [comment]
Then for such case function could be extended to bool creature, bool gob.

**Comment by ForesterDev** [comment]
> Generally this is a step in the right direction but I wonder if its correct for all scripts
> 
> example: script S casts spell X that hits creatures and gameobjects and has SpellHitTarget (old) that does something (to itself, not to targeted creature, think commented out target argument) if that spell hit a creature. Now after applying this change, it will also fire when hitting the gameobject

I agree, but also think, that scripts should foresee this themselves.

**Comment by Shauren** [comment]
> I agree, but also think, that scripts should foresee this themselves.

But the problem here is that you are changing existing scripts written when gameobject caster/target in that specific hook was impossible, so this should be verified by the person doing this change

I don't think this will be an issue for `SpellHit`, only `SpellHitTarget` and most of these spells don't hit gameobjects anyway (and obviously the ones doing explicit ToUnit/ToCreature casts can be counted as verified already)

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_general_vezax.cpp [hunk 2]
```diff
@@ -219,9 +219,9 @@ class boss_general_vezax : public CreatureScript
                 DoMeleeAttackIfReady();
             }
 
-            void SpellHitTarget(Unit* who, SpellInfo const* spell) override
+            void SpellHitTarget(WorldObject* target, SpellInfo const* spellInfo) override
```
Comment (ForesterDev) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_general_vezax.cpp:222]: Ok

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_freya.cpp [hunk 3]
```diff
@@ -1505,12 +1505,16 @@ class npc_unstable_sun_beam : public CreatureScript
                     despawnTimer -= diff;
             }
 
-            void SpellHitTarget(Unit* target, SpellInfo const* spell) override
+            void SpellHitTarget(WorldObject* target, SpellInfo const* spellInfo) override
```
Comment (ForesterDev) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_freya.cpp:1508]: Ok

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_flame_leviathan.cpp [hunk 4]
```diff
@@ -476,16 +476,20 @@ class boss_flame_leviathan : public CreatureScript
                 DoBatteringRamIfReady();
             }
 
-            void SpellHitTarget(Unit* target, SpellInfo const* spell) override
+            void SpellHitTarget(WorldObject* target, SpellInfo const* spellInfo) override
```
Comment (ForesterDev) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_flame_leviathan.cpp:479]: Ok

File: src/server/scripts/Northrend/Nexus/Oculus/boss_varos.cpp [hunk 5]
```diff
@@ -203,9 +203,9 @@ class npc_azure_ring_captain : public CreatureScript
                 me->SetReactState(REACT_AGGRESSIVE);
             }
 
-            void SpellHitTarget(Unit* target, SpellInfo const* spell) override
+            void SpellHitTarget(WorldObject* target, SpellInfo const* spellInfo) override
```
Comment (ForesterDev) [inline src/server/scripts/Northrend/Nexus/Oculus/boss_varos.cpp:206]: Ok SPELL_ICE_BEAM has TARGET_UNIT_SRC_AREA_ENTRY

File: src/server/scripts/Northrend/Naxxramas/boss_sapphiron.cpp [hunk 6]
```diff
@@ -187,12 +187,16 @@ class boss_sapphiron : public CreatureScript
                 EnterPhaseGround(true);
             }
 
-            void SpellHitTarget(Unit* target, SpellInfo const* spell) override
+            void SpellHitTarget(WorldObject* target, SpellInfo const* spellInfo) override
```
Comment (ForesterDev) [inline src/server/scripts/Northrend/Naxxramas/boss_sapphiron.cpp:190]: Ok

File: src/server/scripts/Northrend/Naxxramas/boss_grobbulus.cpp [hunk 7]
```diff
@@ -63,9 +63,9 @@ class boss_grobbulus : public CreatureScript
                 events.ScheduleEvent(EVENT_BERSERK, 12min);
             }
 
-            void SpellHitTarget(Unit* target, SpellInfo const* spell) override
+            void SpellHitTarget(WorldObject* target, SpellInfo const* spellInfo) override
```
Comment (ForesterDev) [inline src/server/scripts/Northrend/Naxxramas/boss_grobbulus.cpp:66]: Ok, SPELL_SLIME_SPRAY has conditions with TARGET_UNIT

File: src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp [hunk 8]
```diff
@@ -671,9 +671,9 @@ class boss_four_horsemen_sir : public CreatureScript
                 }
             }
 
-            void SpellHitTarget(Unit* /*target*/, SpellInfo const* spell) override
+            void SpellHitTarget(WorldObject* /*target*/, SpellInfo const* spellInfo) override
```
Comment (ForesterDev) [inline src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp:674]: Ok, SPELL_HOLY_WRATH has TARGET_UNIT_TARGET_ENEMY

File: src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp [hunk 9]
```diff
@@ -519,9 +519,9 @@ class boss_four_horsemen_thane : public CreatureScript
                 DoMeleeAttackIfReady();
             }
 
-            void SpellHitTarget(Unit* /*target*/, SpellInfo const* spell) override
+            void SpellHitTarget(WorldObject* /*target*/, SpellInfo const* spellInfo) override
```
Comment (ForesterDev) [inline src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp:522]: Ok, SPELL_METEOR has TARGET_UNIT_TARGET_ENEMY

File: src/server/scripts/Northrend/IcecrownCitadel/boss_professor_putricide.cpp [hunk 10]
```diff
@@ -728,9 +728,9 @@ class npc_putricide_oozeAI : public ScriptedAI
         npc_putricide_oozeAI(Creature* creature, uint32 auraSpellId, uint32 hitTargetSpellId) : ScriptedAI(creature),
             _auraSpellId(auraSpellId), _hitTargetSpellId(hitTargetSpellId), _newTargetSelectTimer(0), _instance(creature->GetInstanceScript()) { }
 
-        void SpellHitTarget(Unit* /*target*/, SpellInfo const* spell) override
+        void SpellHitTarget(WorldObject* /*target*/, SpellInfo const* spellInfo) override
```
Comment (ForesterDev) [inline src/server/scripts/Northrend/IcecrownCitadel/boss_professor_putricide.cpp:731]: Ok, _hitTargetSpellId has TARGET_UNIT

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_yogg_saron.cpp [hunk 11]
```diff
@@ -737,12 +737,12 @@ class boss_sara : public CreatureScript
                 }
             }
 
-            void SpellHitTarget(Unit* /*target*/, SpellInfo const* spell) override
+            void SpellHitTarget(WorldObject* /*target*/, SpellInfo const* spellInfo) override
             {
```
Comment (ForesterDev) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_yogg_saron.cpp:741]: Ok, spells have TARGET_UNIT_TARGET_ANY

File: src/server/scripts/Outland/GruulsLair/boss_gruul.cpp [hunk 12]
```diff
@@ -122,11 +122,11 @@ class boss_gruul : public CreatureScript
                 Talk(SAY_DEATH);
             }
 
-            void SpellHitTarget(Unit* target, SpellInfo const* pSpell) override
+            void SpellHitTarget(WorldObject* target, SpellInfo const* spellInfo) override
```
Comment (ForesterDev) [inline src/server/scripts/Outland/GruulsLair/boss_gruul.cpp:125]: Ok

**Comment by ForesterDev** [comment]
@Shauren , I've checked all SpellHitTarget methods in scripts:
1. Those that have TypeId checks - are all OK
2. In those who do not have TypeId checks, I've checked Spells. If spell has target type with UNIT_TARGET or has conditions with UNIT target - then OK.

Only 1 spell has two target types 68981 in Lich King script: TARGET_UNIT_CASTER and TARGET_GAMEOBJECT_SRC_AREA. So I moved this checks to SpellHit to avoid dublicate call for gameobjects.

**Comment by jackpoz** [comment]
could you please fix the merge conflicts of this pr ?

**Comment by jackpoz** [comment]
circle ci reported a missing header in boss_flame_leviathan.cpp, probably Unit.h

**Comment by Shauren** [comment]
its not a missing header, its missing a ToUnit() cast

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_flame_leviathan.cpp [hunk 13]
```diff
@@ -476,15 +476,19 @@ class boss_flame_leviathan : public CreatureScript
...
-                if (spell->Id != SPELL_PURSUED)
+                Unit* unitTarget = target->ToUnit();
+                if (!unitTarget)
+                    return;
+
+                if (spellInfo->Id != SPELL_PURSUED)
                     return;
 
                 _pursueTarget = target->GetGUID();
                 AttackStart(target);
```
Comment (Shauren) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_flame_leviathan.cpp:489]: should change to unitTarget here

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Pull request merged**